frontend/src/utils/fieldValidation.ts

```diff
@@ -206,34 +207,63 @@ export const createMobileValidationRules: ValidationRuleFnEmailAndMobile<
 export const createNumberValidationRules: ValidationRuleFn<NumberFieldBase> = (
   schema,
 ): RegisterOptions => {
-  const { selectedValidation, customVal } = schema.ValidationOptions
+  const { selectedValidation } = schema.ValidationOptions
+  const { selectedLengthValidation, customVal } =
+    schema.ValidationOptions.LengthValidationOptions
+  const { customMin, customMax } =
+    schema.ValidationOptions.RangeValidationOptions
 
   return {
     validate: {
       required: requiredSingleAnswerValidationFn(schema),
-      validNumber: (val?: string) => {
-        if (!val || !customVal) return true
+      validNumberLength: (val: string) => {
+        if (
+          selectedValidation !== NumberSelectedValidation.Length ||
+          !val ||
+          !customVal
+        )
+          return true
 
         const currLen = val.trim().length
 
-        switch (selectedValidation) {
-          case NumberSelectedValidation.Exact:
+        switch (selectedLengthValidation) {
+          case NumberSelectedLengthValidation.Exact:
             return (
               currLen === customVal ||
               simplur`Please enter ${customVal} digit[|s] (${currLen}/${customVal})`
             )
-          case NumberSelectedValidation.Min:
+          case NumberSelectedLengthValidation.Min:
             return (
               currLen >= customVal ||
               simplur`Please enter at least ${customVal} digit[|s] (${currLen}/${customVal})`
             )
-          case NumberSelectedValidation.Max:
+          case NumberSelectedLengthValidation.Max:
             return (
               currLen <= customVal ||
               simplur`Please enter at most ${customVal} digit[|s] (${currLen}/${customVal})`
             )
         }
       },
+      validNumberRange: (val: string) => {
+        if (selectedValidation !== NumberSelectedValidation.Range || !val)
+          return true
+
+        const numVal = Number(val)
```

Comment:
This may not be necessary, as the frontend already ensures that `val` would be a numeric input with `val.replace(/\D/g, ''))`. In comparison, a NaN check is necessary in `createDecimalValidationRules` since `val` may not be a valid numeric input.

Of course, we could also do the check for defensive programming, let me know what you prefer!

Reply:
Yes, we should not rely on guarantees from other upstream functions.

We want to write modular, extensible and maintainable code. It's hard to know when this function might eventually be reused later on. Relying purely on the input types and writing a function handling all cases is preferable to putting a non-static precondition on the function which relies on the user to verify that the precondition is met. Especially in this case where the unexpected case handling is basically only one line. (There might be a tradeoff if the handling is extremely complex - but it isn't in this case.)

Also, yes, I agree there should be one in `createDecimalValidationRules`... Feel free to open a ticket or PR for this if you'd like! But let's not squeeze it into this PR since it's out of scope.